src/analyses/basePriv.ml

```diff
@@ -348,6 +352,188 @@ struct
+    let _,lmust,l = st.priv in
+    let lm = LLock.global x in
+    let tmp = get_mutex_global_g_with_mutex_inits (not (LMust.mem lm lmust)) ask getg x in
+    let local_m = BatOption.default (CPA.bot ()) (L.find_opt lm l) in
+    if is_unprotected ask x then
+      (* We can not rely upon the old value here, it may be too small due to reuse at widening points (and or nice bot/top confusion) in Base *)
+      CPA.find x (CPA.join tmp local_m)
+    else
+      CPA.find x (long_meet st.cpa (CPA.join tmp local_m))
+
+  let read_global ask getg st x =
+    let v = read_global ask getg st x in
+    if M.tracing then M.tracel "priv" "READ GLOBAL %a %B %a = %a\n" d_varinfo x (is_unprotected ask x) CPA.pretty st.cpa VD.pretty v;
+    v
+
+  let write_global ?(invariant=false) ask getg sideg (st: BaseComponents (D).t) x v =
+    let w,lmust,l = st.priv in
+    let lm = LLock.global x in
+    let cpa' =
+      if is_unprotected ask x then
+        st.cpa
+      else
+        CPA.add x v st.cpa
+    in
+    if M.tracing then M.tracel "priv" "WRITE GLOBAL SIDE %a = %a\n" d_varinfo x VD.pretty v;
+    let tid = ThreadId.get_current ask in
+    let sidev = GMutex.singleton tid (CPA.singleton x v) in
+    let l' = L.add lm (CPA.singleton x v) l in
+    sideg (V.global x) (G.create_global sidev);
+    {st with cpa = cpa'; priv = (W.add x w,LMust.add lm lmust,l')}
+
+  let lock (ask: Queries.ask) getg (st: BaseComponents (D).t) m =
+    if Locksets.(not (Lockset.mem m (current_lockset ask))) then (
+      let _,lmust,l = st.priv in
+      let lm = LLock.mutex m in
+      let get_m = get_m_with_mutex_inits (not (LMust.mem lm lmust)) ask getg m in
+      let local_m = BatOption.default (CPA.bot ()) (L.find_opt lm l) in
+      let is_in_Gm x _ = is_protected_by ask m x in
+      let local_m = CPA.filter is_in_Gm local_m in
+      let meet = long_meet st.cpa (CPA.join get_m local_m) in
+      {st with cpa = meet}
+    )
+    else
+      st (* sound w.r.t. recursive lock *)
+
+  let unlock ask getg sideg (st: BaseComponents (D).t) m =
+    let w,lmust,l = st.priv in
+    let cpa' = CPA.fold (fun x v cpa ->
+        if is_protected_by ask m x && is_unprotected_without ask x m then
+          CPA.remove x cpa
+        else
+          cpa
+      ) st.cpa st.cpa
+    in
+    let w' = W.filter (fun v -> not (is_unprotected_without ask v m)) w in
+    let side_needed = W.exists (fun v -> is_protected_by ask m v) w in
+    if not side_needed then
+      {st with cpa = cpa'; priv = (w',lmust,l)}
+    else
+      let is_in_Gm x _ = is_protected_by ask m x in
+      let tid = ThreadId.get_current ask in
+      let sidev = GMutex.singleton tid (CPA.filter is_in_Gm st.cpa) in
+      sideg (V.mutex m) (G.create_mutex sidev);
+      let lm = LLock.mutex m in
+      let l' = L.add lm (CPA.filter is_in_Gm st.cpa) l in
+      {st with cpa = cpa'; priv = (w',LMust.add lm lmust,l')}
+
+  let thread_join (ask:Q.ask) getg exp (st: BaseComponents (D).t) =
```

Comment:
Maybe this should handle the forceful assume-join case as well, like Apron's.

Reply:
Good point! I added it!